utils/kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/kafka/source/delayed/DelayedKafkaSourceFactory.scala

```diff
@@ -20,12 +20,30 @@ object DelayedKafkaSourceFactory {
   final val TimestampFieldParamName = "timestampField"
 
   // TODO: consider changing to lazy parameter and add the same parameter also in "not delayed" kafka sources
-  final val TimestampFieldParameter = Parameter
+  final val fallbackTimestampFieldParameter = Parameter
     .optional(TimestampFieldParamName, Typed[String])
     .copy(
       editor = Some(DualParameterEditor(simpleEditor = StringParameterEditor, defaultMode = DualEditorMode.RAW))
     )
 
+  def timestampFieldParameter(typingResult: Option[TypingResult]): Parameter = {
+    val editorOpt = typingResult
+      .collect { case TypedObjectTypingResult(fields, _, _) => fields.toList }
+      .map(_.collect {
+        case (paramName, typing) if TimestampUtils.supportedTimestampTypes.contains(typing) =>
+          FixedExpressionValue(s"'${paramName}'", paramName)
+      })
+      .filter(_.nonEmpty)
+      .map(_.sortBy(_.label))
+      .map(FixedValuesParameterEditor(_))
+      .map(DualParameterEditor(_, DualEditorMode.SIMPLE))
+      .orElse(Some(DualParameterEditor(simpleEditor = StringParameterEditor, defaultMode = DualEditorMode.RAW)))
```

Comment:
Would be nice to see test showing that even if we give some fields to the user or no field has matching type, user can provide his/her own field. If you think that it is too much and it is better to focus on the final solution it's fine to me to mention in the proposed TODO that we picked this compromise instead.

Reply:
same here